File: cracking_the_coding_interview_book/chapter_4/validate_bst.cpp
#include <iostream>
#include <algorithm>
#include <climits>

using namespace std;

class BinarySearchTree {
    private:
        BinarySearchTree* most_left() {
            if (left == NULL) {
                return this;
            } else {
                return left->most_left();
            }
        }

        BinarySearchTree* most_right() {
            if (right == NULL) {
                return this;
            } else {
                return right->most_right();
            }
        }

        BinarySearchTree* get_pred(BinarySearchTree* node) {
            if (right == node || left == node) return this;
            else if (node->data < data) {
                node->left->get_pred(node);
            } else {
                node->right->get_pred(node);
            }
        }

        void remove_node_link(BinarySearchTree* node) {
            BinarySearchTree* pred = get_pred(node);
            if (pred->left == node) {
                pred->left = NULL;
            } else {
                pred->right = NULL;
            }
            delete node;
        }

    public:
        int data;
        BinarySearchTree* left;
        BinarySearchTree* right;

        BinarySearchTree(int value) {
            data = value;
            left = NULL;
            right = NULL;
        }

        BinarySearchTree* find(int value) {
            if (data == value) {
                return this;
            } else if (value < data && left != NULL) {
                return left->find(value);
            } else if (data < value && right != NULL) {
                return right->find(value);
            } else {
                return NULL;
            }
        }

        void insert(int value) {
            if (value < data) {
                if (left == NULL) {
                    left = new BinarySearchTree(value);
                } else {
                    left->insert(value);
                }
            } else if (data < value) {
                if (right == NULL) {
                    right = new BinarySearchTree(value);
                } else {
                    right->insert(value);
                }
            }
        }

        void print_tree(int h) {
            for (int i = 0; i < h; i++) {
                cout << "   ";
            }
            cout << data << endl;
            if (left != NULL) {
                left->print_tree(h + 1);
            }
            if (right != NULL) {
                right->print_tree(h + 1);
            }
        }

};

bool is_bst(BinarySearchTree* root, int minimal, int maximal) {
    if (root == NULL)
        return true;
    if (root->data < minimal || maximal < root->data)
        return false;
    return is_bst(root->left, minimal, root->data - 1) & is_bst(root->right, root->data + 1, maximal);
}

void insert(BinarySearchTree* bst, int data) {
    bst->insert(data);
    cout << "Insert: " << data << endl;
}

void validate(BinarySearchTree* bst) {
    int h = 0;
    cout << "Is BST: " << is_bst(bst, INT_MIN, INT_MAX) << endl;
}

int main() {
    BinarySearchTree bst(10);
    cout << "INIT: " << 10 << endl;
    insert(&bst, 2);
    insert(&bst, 3);
    insert(&bst, 4);
    insert(&bst, 1);
    insert(&bst, 21);
    insert(&bst, 12);
    insert(&bst, 14);
    insert(&bst, 40);
    insert(&bst, 50);
    cout << endl << "PRINT" << endl;
    bst.print_tree(0);
    validate(&bst);
    bst.data = 800;
    cout << endl << "PRINT" << endl;
    bst.print_tree(0);
    validate(&bst);
    return 0;
}

File: ctci/chapter4/validate_bst.py
#!/bin/python

class Tree:
    def __init__(self, x):
        self.value = x
        self.parent = None
        self.left = None
        self.right = None

    def print_tree(self, padding = 0):
        padding_string = "".join([" "] * padding)
        print "{0}{1}".format(padding_string, self.value)
        if self.left != None:
            self.left.print_tree(padding + 1)
        if self.right != None:
            self.right.print_tree(padding + 1)

def is_bst(root, minimal = 0, maximal = 100):
    # Base case (empty tree)
    if root == None:
        return True

    # Validate current node
    if root.value <= minimal:
        return False
    if maximal <= root.value:
        return False

    # Validate left
    if not is_bst(root.left, minimal=minimal, maximal=root.value):
        return False

    # Validate right
    if not is_bst(root.right, minimal=root.value, maximal=maximal):
        return False

    return True

def test(root):
    print "============================================"
    root.print_tree()
    print "Result: {0}".format(is_bst(root))


root = Tree(5)
test(root)

root.left = Tree(2)
root.right = Tree(8)
test(root)

root.right = None
test(root)

root.right = Tree(8)
root.right.left = Tree(7)
test(root)

root.right = Tree(8)
root.right.left = Tree(3)
test(root)
